### PR TITLE
Manage git repositories via CRDs

### PR DESCRIPTION
**Pull request by Kidswiss**
This commit contains the implementation for Gitlab to manage the repos.

### CODE REVIEW

**Comment by Kidswiss** [comment]
@srueg I renamed `bootstrapTokenValid` in the cluster status to `tokenValid` is that okay for you?

**Comment by Kidswiss** [comment]
The git functionality should now be okay for a first version.

Also a proposal:

Let's use golang's ldflags to automagically set the version and build date during CI/CD
```
 go run -ldflags="-X 'main.version=$(git describe --tags)' -X 'main.date=$(date)'" cmd/manager/main.go

# in main.go
// set these with ldflags
var version = "dev"
var date = "now"

func printVersion() {
	log.Info(fmt.Sprintf("Operator Version: %s", version))
	log.Info(fmt.Sprintf("Build Date: %s", date))
*SNIP*
```

**Comment by srueg** [comment]
> The git functionality should now be okay for a first version.
> 
> Also a proposal:
> 
> Let's use golang's ldflags to automagically set the version and build date during CI/CD
> 
> ```
> # set version during build/run
> go build/run -ldflags="-X 'main.version=$(git describe --tags)' -X 'main.date=$(date)'" cmd/manager/main.go
> 
> # in main.go
> // set these with ldflags
> var version = "dev"
> var date = "now"
> 
> func printVersion() {
> 	log.Info(fmt.Sprintf("Operator Version: %s", version))
> 	log.Info(fmt.Sprintf("Build Date: %s", date))
> *SNIP*
> ```

:+1: Implemented in https://github.com/projectsyn/lieutenant-operator/pull/5

File: pkg/git/gitlab/gitlab.go [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+	for k, v := range localKeys {
+		mergedKey := v.Type + " " + v.Key
+		keyOpts := &gitlab.AddDeployKeyOptions{
+			Title:   &k,
+			Key:     &mergedKey,
+			CanPush: &v.WriteAccess,
+		}
+
+		if force {
+			g.overwriteKey(existingKeys, k)
```
Comment (srueg) [inline pkg/git/gitlab/gitlab.go]: Maybe we should delete the old key only once adding the new one was successful?

File: pkg/git/manager/manager.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+	Read() error
+	Delete() error
+	Connect() error
+}
+
+// Implementation is a set of functions needed to get the right git implementation
+// for the given URL.
+type Implementation interface {
+	// IsType returns true, if the given URL is handleable by the given implementation (Github,Gitlab, etc.)
+	IsType(URL string) bool
```
Comment (srueg) [inline pkg/git/manager/manager.go]: Maybe we should differentiate between false and error, to make debugging easier?
I.e. if a connection error occurs, it'll now look like it couldn't find a proper implementation...

**Comment by Kidswiss** [comment]
@sruegg ready for another look 👍

File: pkg/controller/gitrepo/gitrepo_reconcile.go [hunk 3]
```diff
@@ -18,18 +26,100 @@ func (r *ReconcileGitRepo) Reconcile(request reconcile.Request) (reconcile.Resul
...
+	gitRepo.Status.Phase = &phaseToSet
+
+	if phaseToSet == synv1alpha1.Created {
+		gitRepo.Status.URL = repoImpl.FullURL().String()
+	}
+
+	return r.client.Status().Update(context.TODO(), gitRepo)
+}
+
+func (r *ReconcileGitRepo) fetchInstance(instance *synv1alpha1.GitRepo, namespacedName types.NamespacedName) error {
```
Comment (srueg) [inline pkg/controller/gitrepo/gitrepo_reconcile.go]: This method can be removed, since only used once

File: pkg/helpers/crd.go [hunk 4]
```diff
@@ -23,11 +23,13 @@ func CreateGitRepo(obj metav1.Object, gvk schema.GroupVersionKind, template *syn
 
 	tenantName := obj.GetName()
 	tenantNamespace := obj.GetNamespace()
-	if tenantRef != nil {
+	if tenantRef != nil || tenantName == "" {
 		tenantName = tenantRef.Name
```
Comment (srueg) [inline pkg/helpers/crd.go]: `tenantRef` could be `nil` here. Should probably be AND instead of OR?

**Comment by srueg** [comment]
For the `GitRepoSpec`, could we just embed the `GitRepoTemplate` instead of having a property with the name? It would reduce the nesting (by removing `gitRepoTemplate:`).

**Comment by Kidswiss** [comment]
> For the `GitRepoSpec`, could we just embed the `GitRepoTemplate` instead of having a property with the name? It would reduce the nesting (by removing `gitRepoTemplate:`).

It's already implemented like this:
```
// GitRepoSpec defines the desired state of GitRepo
type GitRepoSpec struct {
	GitRepoTemplate `json:"gitRepoTemplate"`
	// TenantRef references the tenant this repo belongs to
	TenantRef corev1.LocalObjectReference `json:"tenantRef"`
}
```

Although I had to change the json tag from `json:"gitRepoTemplate"` to `json:",inline"` to make the code generator happy again, is it requires a valis json tag for every field. Now the additional nesting isn't necessary anymore.

File: pkg/helpers/crd.go (hunk 4 above)
Comment (Kidswiss) [reply pkg/helpers/crd.go]: Simplified the logic.

File: pkg/apis/syn/v1alpha1/gitrepo_types.go [hunk 5]
```diff
@@ -63,31 +55,21 @@ type DeployKey struct {
...
-	Type               string      `json:"type,omitempty"`
-}
-
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
 // GitRepo is the Schema for the gitrepos API
 // +kubebuilder:subresource:status
 // +kubebuilder:resource:path=gitrepos,scope=Namespaced
-// +kubebuilder:printcolumn:name="Repo Name",type="string",JSONPath=".spec.repoName"
+// +kubebuilder:printcolumn:name="Repo Name",type="string",JSONPath=".spec.gitRepoTemplate.repoName"
```
Comment (srueg) [inline pkg/apis/syn/v1alpha1/gitrepo_types.go]:
```suggestion
// +kubebuilder:printcolumn:name="Repo Name",type="string",JSONPath=".spec.repoName"
```

**Pull request merged**